src/compile/compile.ts

```diff
@@ -65,23 +69,43 @@ function getTopLevelProperties(topLevelSpec: TopLevel<any>, config: Config) {
  * Note: this couldn't be `model.assemble()` since the top-level model
  * needs some special treatment to generate top-level properties.
  */
-function assembleTopLevelModel(model: Model, topLevelProperties: TopLevelProperties) {
+function assembleTopLevelModel(model: Model, topLevelProperties: TopLevelProperties & LayoutSizeMixins) {
   // TODO: change type to become VgSpec
 
   // Config with Vega-Lite only config removed.
   const vgConfig = model.config ? stripAndRedirectConfig(model.config) : undefined;
 
   // autoResize has to be put under autosize
-  const {autoResize, ...topLevelProps} = topLevelProperties;
+  const {autosize, ...topLevelProps} = topLevelProperties;
 
   const title = model.assembleTitle();
   const style = model.assembleGroupStyle();
 
+  let layoutSignals = model.assembleLayoutSignals();
+
+  // move width and height signals with values to top level
+  layoutSignals = layoutSignals.filter(signal => {
+    if ((signal.name === 'width' || signal.name === 'height') && signal.value !== undefined) {
+      topLevelProps[signal.name] = signal.value;
+      return false;
+    }
+    return true;
+  });
+
+  if (autosize === 'fit' || (autosize && isAutosize(autosize) && autosize.type === 'fit')) {
+    if (!topLevelProps.width || !topLevelProps.height) {
```

Comment:
Instead of warning, I think we better modify scale range logic to support this case.

Reply:
I'll do both. We should support this case but also show a warning that we don't determine the size in a data driven way.

Reply:
I guess we should only warn when users explicitly specify both autosize and range step?

(Otherwise, why would we need to warn?)

Reply:
Because the behavior is different from when autosize is not fit